### ISSUE TITLE
Destructuring `this.get()`, `this.get({ virtual: true })` or `this.get({ virtual: false })`

### ISSUE THREAD

**Issue by giovannipiller** [ractivejs/ractive#3116]
# [update] - 25th Oct 2017
Mystery solved, [take a look here](https://github.com/ractivejs/ractive/issues/3116#issuecomment-339263367).

Issue was previously titled: "**Inconsistencies between `this.root.get()` and `this.root.get({ virtual: true })`**"

---

## Description:
According to the docs, `this.root.get()` and `this.root.get({ virtual: true })` should be pretty much the same:

> virtual (boolean): When set to false, excludes virtual keypaths (computations, links, mappings, etc.). Defaults to true for the root keypath and false for keypaths other than the root.

That being said, I noticed that `this.root.get({ virtual: true })` seems to behave somewhat differently.

## Versions affected:
0.9.5

## Platforms affected:
tested on Chrome

## Reproduction:
JSfiddle: https://jsfiddle.net/kfguw0a4/2/

```javascript
const r = window.r = new Ractive({
  el: '#main',
  template: '#template',
  data: {
    myobject: {
      test: true
    },
  },
  oninit() {
    console.clear();

    // we're in the root component
    if (this.root === this) {
      {
        const { myobject } = this.get();

        console.warn(myobject === this.get('myobject')); // => true
      }

      {
        const { myobject } = this.get({ virtual: true });

        console.info(myobject === this.get('myobject')); // => false
      }

      console.info('Both console messages should be the same, as the default for this.get() is virtual: true, from root components');
      console.info('see: https://ractive.js.org/api/#ractiveget');
    }
  }
});
```

**Comment by giovannipiller** [comment]
# Additional thoughts

I always thought that some form of `const { myobject } = this.get()` returned a reference to `myobject`, but it looks like this is pretty much never the case.

Built a fiddle to give you an idea: https://jsfiddle.net/c4yqvzue/2/

```javascript
const MyComponent = Ractive.extend({
  isolated: true,

  oninit() {
    this.fire('myevent');
  }
});

// if you change your console context to "result", you can interact with
// your Ractive instance with the glboal "r"
const r = window.r = new Ractive({
  el: '#main',
  template: '<MyComponent myobject="{{myobject}}"/>',
  data: {
    myobject: {
      test: true
    },
  },

  components: {
    MyComponent
  },

  on: {
    // fired immediately as MyComponent has oninit
    'MyComponent.myevent': function({
      component
    }) {
      console.clear();
      
      const root = this;

      // Checking .get() from `component`
      {
        const { myobject } = component.get();

        // should be `true, true`
        console.warn(myobject === component.get('myobject'), myobject === root.get('myobject')); // => false, false
      }

      {
        const { myobject } = component.get({ virtual: true });

        // should be `true, true`
        console.warn(myobject === component.get('myobject'), myobject === root.get('myobject')); // => false, false
      }

      {
        const { myobject } = component.get({ virtual: false });

        // should be `true, true`
        console.warn(myobject === component.get('myobject'), myobject === root.get('myobject')); // => false, false
      }
      
      // Checking .get() from `root`
      {
        const { myobject } = root.get();

        // should be `true, true`
        console.warn(myobject === component.get('myobject'), myobject === root.get('myobject')); // => false, false
      }

      {
        const { myobject } = root.get({ virtual: true });

        // should be `true, true`
        console.warn(myobject === component.get('myobject'), myobject === root.get('myobject')); // => false, false
      }

      {
        const { myobject } = root.get({ virtual: false });

        // WORKS FINE!!
        console.log('this one is fine', myobject === component.get('myobject'), myobject === root.get('myobject')); // => true, true
      }
      
      // Checking .get('myobject') from `component` and `root`
      {
      	console.info("Checking .get('myobject') from `component` and `root`", component.get('myobject') === root.get('myobject')); // => true, as expected
      }
    }
  }
});
```

**Comment by evs-chris** [comment]
So this is one of those weird corner cases that pops up because observing the output also changes it. A virtual get is a hybrid between actual object values and the models that ractive is tracking. If you have `r.set('foo', { bar, baz })` and neither `foo.bar` nor `foo.baz` are accessed by ractive, it won't have models for them. When there is no model, the virtual get just returns the value, leaving you with the original `bar` and `baz`.

As soon as you `r.get('foo.bar')`, use it in template, or otherwise cause it to be retrieved, ractive creates a model for `foo.bar`, as that keypath is now in scope for the instance and may be required elsewhere for a number of reasons, including capture. When there _is_ a model for a keypath, it is virtualized as well, so you'll get a shallow copy of `foo.bar` from that point on.

TL;DR if ractive has a model for the keypath, it gets virtualized; otherwise, you just get the value directly.

Does that mostly make sense?

**Comment by giovannipiller** [comment]
Hmmm..

Not sure I got this right, but it doesn't look like accessing `this.get('myobject')` has an impact on my code: https://jsfiddle.net/c4yqvzue/4/

The issue I experienced on my code was something like this: https://jsfiddle.net/rp8am4z4/1/
I was trying to use destructuring to access an object, change it, and then perform an `this.update()`.

This obviously fails because the objects returned are not a reference to the actual object in most situations.

---

What about https://github.com/ractivejs/ractive/issues/3116#issue-267625524 ?
Shouldn't `this.root.get()` equal `this.root.get({ virtual: true })`?

**Comment by evs-chris** [comment]
I'll look at the fiddle when I get to a real machine, but if you want to use `update`, you'd want

```js
const { myobject } = r.get({ virtual: false });
myobject.newValue = true;
r.update('myobject');
```

because a non-virtual get returns the actual underlying values every time.

**Comment by evs-chris** [comment]
Ah! It took me a bit to wrap my head around the example.

You have a linked/mapped path in the component that you're trying to modify using destructuring. The problem is that you don't get direct access to the value if it's a link, and changing the virtual value returned from `component.get()` or `component.get('path', { virtual: true })` won't actually change any of the object tree underlying the instance's data.

Unfortunately, the only solution to that is to use the keypath directly e.g. `const myobject = r.get('myobject');` which will give you back the actual value. If you get a non-virtual result, then the linked data won't be present - because it's not present in the actual value - the source instance owns that data. If you need to be able to use destructuring with linked data, you could link in a wrapper object e.g. `<cmp stuff="{{wrapper}}" />` and use `const { myobject } = this.get('stuff');` where the source `wrapper` is an object with `myobject` inside.

**Comment by giovannipiller** [comment]
Ohhh, ok I think I kinda get it now.

> As soon as you r.get('foo.bar'), use it in template, or otherwise cause it to be retrieved, ractive creates a model for foo.bar, as that keypath is now in scope for the instance and may be required elsewhere for a number of reasons, including capture. When there is a model for a keypath, it is virtualized as well, so you'll get a shallow copy of foo.bar from that point on.

For future reference, let me write down an example.

### The following allows changes after destructuring

```javascript
const { myobject } = this.get();
myobject.example = 'edit object'

console.log(myobject === this.get('myobject')) // => true
console.log(this.get('myobject').example) // => 'edit object'
```

### HOWEVER, if you access `myobject` in any way before executing it

```javascript
this.get('myobject'); // here, or even from the template

const { myobject } = this.get();
myobject.example = 'edit object'

console.log(myobject === this.get('myobject')) // => false
console.log(this.get('myobject').example) // => undefined
```

# TL;DR
To avoid issues:

**Don't destructure a `this.get()` or `this.get({ virtual: true })` as they will likely return a shallow copy.**
Use `this.get({ virtual: false })` to do it, but be aware that **it will only work from the component that owns `myobject`**.

```javascript
this.get('myobject');

const { myobject } = this.get({ virtual: false });
myobject.example = 'edit object'

console.log(myobject === this.get('myobject')) // => true
console.log(this.get('myobject').example) // => 'edit object'
```
